From c11f47481a48113588a4f942d960a8d17190cd55 Mon Sep 17 00:00:00 2001
From: Olivia Lee <olivia.lee@collabora.com>
Date: Fri, 8 Aug 2025 21:00:47 -0700
Subject: [PATCH] panvk: stop CPU mapping all index buffers on JM

This can be removed now that we do compute shader minmax. Fixes failed
mmap errors on PPSSPP and some other vulkan applications:

MESA: error: mmap(..., size=4194304, prot=2, flags=0x1) failed: Invalid argument

Fixes: e25064c026a ("panvk: Use indirect path for indexed draw on JM")
Signed-off-by: Olivia Lee <olivia.lee@collabora.com>
Reviewed-by: Mary Guillemard <mary.guillemard@collabora.com>
Part-of: <https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/36695>
---
 src/panfrost/vulkan/panvk_buffer.c | 35 +-----------------------------
 1 file changed, 1 insertion(+), 34 deletions(-)

diff --git a/src/panfrost/vulkan/panvk_buffer.c b/src/panfrost/vulkan/panvk_buffer.c
index abbaf18e37670..ebbc014ab4ec2 100644
--- a/src/panfrost/vulkan/panvk_buffer.c
+++ b/src/panfrost/vulkan/panvk_buffer.c
@@ -54,12 +54,6 @@ VKAPI_ATTR VkResult VKAPI_CALL
 panvk_BindBufferMemory2(VkDevice _device, uint32_t bindInfoCount,
                         const VkBindBufferMemoryInfo *pBindInfos)
 {
-   VK_FROM_HANDLE(panvk_device, device, _device);
-   const struct panvk_physical_device *phys_dev =
-      to_panvk_physical_device(device->vk.physical);
-   const unsigned arch = pan_arch(phys_dev->kmod.props.gpu_id);
-   VkResult result = VK_SUCCESS;
-
    for (uint32_t i = 0; i < bindInfoCount; i++) {
       VK_FROM_HANDLE(panvk_device_memory, mem, pBindInfos[i].memory);
       VK_FROM_HANDLE(panvk_buffer, buffer, pBindInfos[i].buffer);
@@ -73,35 +67,8 @@ panvk_BindBufferMemory2(VkDevice _device, uint32_t bindInfoCount,
       assert(buffer->vk.device_address == 0);
 
       buffer->vk.device_address = mem->addr.dev + pBindInfos[i].memoryOffset;
-
-      /* FIXME: Only host map for index buffers so we can do the min/max
-       * index retrieval on the CPU. This is all broken anyway and the
-       * min/max search should be done with a compute shader that also
-       * patches the job descriptor accordingly (basically an indirect draw).
-       *
-       * Make sure this goes away as soon as we fixed indirect draws.
-       */
-      if (arch < 9 && (buffer->vk.usage & VK_BUFFER_USAGE_INDEX_BUFFER_BIT)) {
-         VkDeviceSize offset = pBindInfos[i].memoryOffset;
-         VkDeviceSize pgsize = getpagesize();
-         off_t map_start = offset & ~(pgsize - 1);
-         off_t map_end = offset + buffer->vk.size;
-         void *map_addr =
-            pan_kmod_bo_mmap(mem->bo, map_start, map_end - map_start,
-                             PROT_WRITE, MAP_SHARED, NULL);
-
-         if (map_addr == MAP_FAILED) {
-            result = panvk_errorf(device, VK_ERROR_OUT_OF_HOST_MEMORY,
-                                  "Failed to CPU map index buffer");
-            if (bind_status)
-               *bind_status->pResult = result;
-            continue;
-         }
-
-         buffer->host_ptr = map_addr + (offset & pgsize);
-      }
    }
-   return result;
+   return VK_SUCCESS;
 }
 
 VKAPI_ATTR VkResult VKAPI_CALL
-- 
GitLab

From 125de6cd638212913504d0a1ced96ffe9eacbedd Mon Sep 17 00:00:00 2001
From: Boris Brezillon <boris.brezillon@collabora.com>
Date: Fri, 22 Aug 2025 14:04:57 +0200
Subject: [PATCH] panvk: Remove leftovers from CPU-side min/max index
 calculation

Commit c11f47481a48 ("panvk: stop CPU mapping all index buffers on JM")
stopped mapping the buffer objects on v9-, but it forgot to remove
panvk_buffer::host_ptr and panvk_cmd_graphics_state::host_addr.

Signed-off-by: Boris Brezillon <boris.brezillon@collabora.com>
Reviewed-by: Olivia Lee <olivia.lee@collabora.com>
Part-of: <https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/36934>
---
 src/panfrost/vulkan/panvk_buffer.c      | 11 -----------
 src/panfrost/vulkan/panvk_buffer.h      |  9 ---------
 src/panfrost/vulkan/panvk_cmd_draw.h    |  3 ---
 src/panfrost/vulkan/panvk_vX_cmd_draw.c |  7 -------
 4 files changed, 30 deletions(-)

diff --git a/src/panfrost/vulkan/panvk_buffer.c b/src/panfrost/vulkan/panvk_buffer.c
index ebbc014ab4ec2..14779c72be804 100644
--- a/src/panfrost/vulkan/panvk_buffer.c
+++ b/src/panfrost/vulkan/panvk_buffer.c
@@ -103,16 +103,5 @@ panvk_DestroyBuffer(VkDevice _device, VkBuffer _buffer,
    if (!buffer)
       return;
 
-   if (buffer->host_ptr) {
-      VkDeviceSize pgsize = getpagesize();
-      uintptr_t map_start = (uintptr_t)buffer->host_ptr & ~(pgsize - 1);
-      uintptr_t map_end =
-         ALIGN_POT((uintptr_t)buffer->host_ptr + buffer->vk.size, pgsize);
-      ASSERTED int ret = os_munmap((void *)map_start, map_end - map_start);
-
-      assert(!ret);
-      buffer->host_ptr = NULL;
-   }
-
    vk_buffer_destroy(&device->vk, pAllocator, &buffer->vk);
 }
diff --git a/src/panfrost/vulkan/panvk_buffer.h b/src/panfrost/vulkan/panvk_buffer.h
index 1d3096731c2ed..5b3b47f176600 100644
--- a/src/panfrost/vulkan/panvk_buffer.h
+++ b/src/panfrost/vulkan/panvk_buffer.h
@@ -14,15 +14,6 @@ struct panvk_priv_bo;
 
 struct panvk_buffer {
    struct vk_buffer vk;
-
-   /* FIXME: Only used for index buffers to do the min/max index retrieval on
-    * the CPU. This is all broken anyway and the min/max search should be done
-    * with a compute shader that also patches the job descriptor accordingly
-    * (basically an indirect draw).
-    *
-    * Make sure this field goes away as soon as we fixed indirect draws.
-    */
-   void *host_ptr;
 };
 
 VK_DEFINE_NONDISP_HANDLE_CASTS(panvk_buffer, vk.base, VkBuffer,
diff --git a/src/panfrost/vulkan/panvk_cmd_draw.h b/src/panfrost/vulkan/panvk_cmd_draw.h
index b1d7c11edfa9e..d9cb097691563 100644
--- a/src/panfrost/vulkan/panvk_cmd_draw.h
+++ b/src/panfrost/vulkan/panvk_cmd_draw.h
@@ -170,9 +170,6 @@ struct panvk_cmd_graphics_state {
    /* Index buffer */
    struct {
       uint64_t dev_addr;
-#if PAN_ARCH < 9
-      void *host_addr;
-#endif
       uint64_t size;
       uint8_t index_size;
    } ib;
diff --git a/src/panfrost/vulkan/panvk_vX_cmd_draw.c b/src/panfrost/vulkan/panvk_vX_cmd_draw.c
index 735600139ac81..1ec1eabfc68b9 100644
--- a/src/panfrost/vulkan/panvk_vX_cmd_draw.c
+++ b/src/panfrost/vulkan/panvk_vX_cmd_draw.c
@@ -903,10 +903,6 @@ panvk_per_arch(CmdBindIndexBuffer2)(VkCommandBuffer commandBuffer,
       cmdbuf->state.gfx.ib.size = panvk_buffer_range(buf, offset, size);
       assert(cmdbuf->state.gfx.ib.size <= UINT32_MAX);
       cmdbuf->state.gfx.ib.dev_addr = panvk_buffer_gpu_ptr(buf, offset);
-#if PAN_ARCH < 9
-      cmdbuf->state.gfx.ib.host_addr =
-         buf && buf->host_ptr ? buf->host_ptr + offset : NULL;
-#endif
    } else {
       cmdbuf->state.gfx.ib.size = 0;
       /* In case of NullDescriptors, we need to set a non-NULL address and rely
@@ -914,9 +910,6 @@ panvk_per_arch(CmdBindIndexBuffer2)(VkCommandBuffer commandBuffer,
        * that this only works for v10+, as v9 does not have a way to specify the
        * index buffer size. */
       cmdbuf->state.gfx.ib.dev_addr = PAN_ARCH >= 10 ? 0x1000 : 0;
-#if PAN_ARCH < 9
-      cmdbuf->state.gfx.ib.host_addr = 0;
-#endif
    }
    cmdbuf->state.gfx.ib.index_size = vk_index_type_to_bytes(indexType);
 
-- 
GitLab

From a382e97339a350f54d0a27c33905883865d5e43a Mon Sep 17 00:00:00 2001
From: Mary Guillemard <mary.guillemard@collabora.com>
Date: Tue, 24 Jun 2025 18:50:36 +0200
Subject: [PATCH] panvk: Implement indirect draw for Bifrost on JM

Signed-off-by: Mary Guillemard <mary.guillemard@collabora.com>
Reviewed-by: Olivia Lee <olivia.lee@collabora.com>
Part-of: <https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/35724>
---
 src/panfrost/ci/panfrost-g52-fails.txt     |   3 -
 src/panfrost/ci/panfrost-g52-flakes.txt    |   1 +
 src/panfrost/ci/panfrost-g52-skips.txt     |  16 +-
 src/panfrost/vulkan/jm/panvk_vX_cmd_draw.c | 298 ++++++++++++++++++++-
 4 files changed, 300 insertions(+), 18 deletions(-)

diff --git a/src/panfrost/ci/panfrost-g52-fails.txt b/src/panfrost/ci/panfrost-g52-fails.txt
index 5450b390a3bb7..0e0cfc1fcb2cf 100644
--- a/src/panfrost/ci/panfrost-g52-fails.txt
+++ b/src/panfrost/ci/panfrost-g52-fails.txt
@@ -396,9 +396,6 @@ dEQP-VK.api.device_init.create_device_global_priority_khr.basic,Fail
 dEQP-VK.api.device_init.create_device_global_priority_query.basic,Fail
 dEQP-VK.api.device_init.create_device_global_priority_query_khr.basic,Fail
 
-# CmdDrawIndirect not supported yet
-dEQP-VK.api.command_buffers.many_indirect_draws_on_secondary,Crash
-
 dEQP-VK.device_group.afr,Fail
 dEQP-VK.device_group.afr_dedicated,Fail
 dEQP-VK.device_group.afr_sys,Fail
diff --git a/src/panfrost/ci/panfrost-g52-flakes.txt b/src/panfrost/ci/panfrost-g52-flakes.txt
index b0c451e0bec01..fc725ff12fa34 100644
--- a/src/panfrost/ci/panfrost-g52-flakes.txt
+++ b/src/panfrost/ci/panfrost-g52-flakes.txt
@@ -38,6 +38,7 @@ afbcp-glx@glx-multithread-clearbuffer
 dEQP-VK.memory.pipeline_barrier.host_write_vertex_buffer.1048576_vertex_buffer_stride_2
 dEQP-VK.memory.pipeline_barrier.host_write_uniform_buffer.1048576
 dEQP-VK.memory.pipeline_barrier.host_write_uniform_texel_buffer.1048576
+dEQP-VK.memory.pipeline_barrier.host_write_index_buffer.1048576
 
 # Sometime timeout
 dEQP-VK.memory.pipeline_barrier.host_write_storage_buffer.1048576
diff --git a/src/panfrost/ci/panfrost-g52-skips.txt b/src/panfrost/ci/panfrost-g52-skips.txt
index b01f34486a262..da76fc7841161 100644
--- a/src/panfrost/ci/panfrost-g52-skips.txt
+++ b/src/panfrost/ci/panfrost-g52-skips.txt
@@ -73,19 +73,6 @@ shaders@glsl-bug-110796
 
 dEQP-VK.dynamic_rendering.primary_cmd_buff.random.seed*
 
-# indirect draw not supported yet
-dEQP-VK.synchronization.*indirect*
-dEQP-VK.synchronization2.*indirect*
-dEQP-VK.draw.renderpass.indirect_draw.*
-dEQP-VK.draw.renderpass.*.draw_indirect*
-dEQP-VK.draw.renderpass.*.draw_indexed_indirect*
-dEQP-VK.draw.dynamic_rendering.*.draw_indirect*
-dEQP-VK.draw.dynamic_rendering.*.draw_indexed_indirect*
-dEQP-VK.draw.dynamic_rendering.*.indirect_draw*
-dEQP-VK.multiview.draw_indirect*
-dEQP-VK.multiview.dynamic_rendering.draw_indirect*
-dEQP-VK.multiview.renderpass2.draw_indirect*
-
 # VKCTS bug?
 # Jump to 0x0 (XXX: need more research)
 dEQP-VK.glsl.shader_expect_assume.*
@@ -117,6 +104,9 @@ dEQP-VK.compute.pipeline.zero_initialize_workgroup_memory.types.f32mat4x4
 # Job timeout
 dEQP-VK.graphicsfuzz.spv-composites
 
+# Cause OOM when running with alloacation tests
+dEQP-VK.api.command_buffers.many_indirect_draws_on_secondary
+
 # Slow tests (>= 30s)
 dEQP-VK.api.external.fence.sync_fd.export_multiple_times_temporary
 dEQP-VK.api.external.semaphore.sync_fd.export_multiple_times_temporary
diff --git a/src/panfrost/vulkan/jm/panvk_vX_cmd_draw.c b/src/panfrost/vulkan/jm/panvk_vX_cmd_draw.c
index 9e300332ec143..a7494358d4d73 100644
--- a/src/panfrost/vulkan/jm/panvk_vX_cmd_draw.c
+++ b/src/panfrost/vulkan/jm/panvk_vX_cmd_draw.c
@@ -17,6 +17,7 @@
 #include "panvk_cmd_desc_state.h"
 #include "panvk_cmd_draw.h"
 #include "panvk_cmd_meta.h"
+#include "panvk_cmd_precomp.h"
 #include "panvk_device.h"
 #include "panvk_entrypoints.h"
 #include "panvk_image.h"
@@ -1550,6 +1551,264 @@ panvk_cmd_draw(struct panvk_cmd_buffer *cmdbuf, struct panvk_draw_data *draw)
    cmdbuf->state.gfx.vs.previous_draw_was_indirect = false;
 }
 
+static void
+panvk_cmd_draw_indirect(struct panvk_cmd_buffer *cmdbuf,
+                        struct panvk_draw_data *draw)
+{
+   const struct panvk_shader_variant *vs = panvk_shader_hw_variant(cmdbuf->state.gfx.vs.shader);
+   VkResult result;
+
+   /* If there's no vertex shader, we can skip the draw. */
+   if (!panvk_priv_mem_dev_addr(vs->rsd))
+      return;
+
+   /* Needs to be done before get_fs() is called because it depends on
+    * fs.required being initialized. */
+   cmdbuf->state.gfx.fs.required =
+      fs_required(&cmdbuf->state.gfx, &cmdbuf->vk.dynamic_graphics_state);
+
+   result = prepare_draw(cmdbuf, draw);
+   if (result != VK_SUCCESS)
+      return;
+
+   struct panvk_batch *batch = cmdbuf->cur_batch;
+   const struct vk_input_assembly_state *ia =
+      &cmdbuf->vk.dynamic_graphics_state.ia;
+   const struct vk_vertex_input_state *vi =
+      cmdbuf->vk.dynamic_graphics_state.vi;
+
+   unsigned copy_desc_job_id =
+      draw->jobs.vertex_copy_desc.gpu
+         ? pan_jc_add_job(&batch->vtc_jc, MALI_JOB_TYPE_COMPUTE, false, false,
+                          0, 0, &draw->jobs.vertex_copy_desc, false)
+         : 0;
+
+   if (draw->jobs.frag_copy_desc.gpu) {
+      /* We don't need to add frag_copy_desc as a dependency because the
+       * tiler job doesn't execute the fragment shader, the fragment job
+       * will, and the tiler/fragment synchronization happens at the batch
+       * level. */
+      pan_jc_add_job(&batch->vtc_jc, MALI_JOB_TYPE_COMPUTE, false, false, 0, 0,
+                     &draw->jobs.frag_copy_desc, false);
+   }
+
+   uint32_t view_mask = cmdbuf->state.gfx.render.view_mask;
+   assert(view_mask == 0 || util_bitcount(view_mask) <= batch->fb.layer_count);
+   uint32_t enabled_layer_count = view_mask
+                                     ? util_bitcount(view_mask)
+                                     : cmdbuf->state.gfx.render.layer_count;
+   const struct panvk_shader_variant *fs = panvk_shader_only_variant(get_fs(cmdbuf));
+
+   for (uint32_t i = 0; i < enabled_layer_count; i++) {
+      /* Force a new push uniform block to be allocated */
+      gfx_state_set_dirty(cmdbuf, VS_PUSH_UNIFORMS);
+
+      result = panvk_draw_prepare_varyings(cmdbuf, draw);
+      if (result != VK_SUCCESS)
+         return;
+
+      draw->info.layer_id = (view_mask != 0) ? u_bit_scan(&view_mask) : i;
+      if (draw->info.layer_id > 0) {
+         cmdbuf->state.gfx.sysvals.layer_id = draw->info.layer_id;
+         gfx_state_set_dirty(cmdbuf, FS_PUSH_UNIFORMS);
+      }
+
+      result = panvk_per_arch(cmd_prepare_push_uniforms)(
+         cmdbuf, vs, 1);
+      if (result != VK_SUCCESS)
+         return;
+
+      if (fs) {
+         result = panvk_per_arch(cmd_prepare_push_uniforms)(
+            cmdbuf, fs, 1);
+         if (result != VK_SUCCESS)
+            return;
+      }
+
+      result = panvk_draw_prepare_tiler_context(cmdbuf, draw);
+      if (result != VK_SUCCESS)
+         return;
+
+      if (vs->info.vs.idvs) {
+         result = panvk_draw_prepare_idvs_job(cmdbuf, draw);
+
+         if (result != VK_SUCCESS)
+            return;
+      } else {
+         result = panvk_draw_prepare_vertex_job(cmdbuf, draw);
+
+         if (result != VK_SUCCESS)
+            return;
+
+         bool needs_tiling =
+            !cmdbuf->vk.dynamic_graphics_state.rs.rasterizer_discard_enable ||
+            cmdbuf->state.gfx.occlusion_query.mode !=
+               MALI_OCCLUSION_MODE_DISABLED;
+
+         if (needs_tiling) {
+            result = panvk_draw_prepare_tiler_job(cmdbuf, draw);
+
+            if (result != VK_SUCCESS)
+               return;
+         }
+      }
+
+      assert(draw->info.indirect.buffer_dev_addr != 0 || draw->info.index.size);
+
+      uint32_t attrib_bufs_valid = vi->bindings_valid;
+      uint32_t attribs_valid = vi->attributes_valid;
+      uint64_t first_vertex_sysval = 0x8ull << 60;
+      uint64_t first_instance_sysval = 0x8ull << 60;
+      uint64_t raw_vertex_offset_sysval = 0x8ull << 60;
+      if (shader_uses_sysval(vs, graphics, vs.first_vertex)) {
+         first_vertex_sysval = cmdbuf->state.gfx.vs.push_uniforms +
+                               shader_remapped_sysval_offset(
+                                  vs, sysval_offset(graphics, vs.first_vertex));
+      }
+
+      if (shader_uses_sysval(vs, graphics, vs.base_instance)) {
+         first_instance_sysval =
+            cmdbuf->state.gfx.vs.push_uniforms +
+            shader_remapped_sysval_offset(
+               vs, sysval_offset(graphics, vs.base_instance));
+      }
+
+      if (shader_uses_sysval(vs, graphics, vs.raw_vertex_offset)) {
+         raw_vertex_offset_sysval =
+            cmdbuf->state.gfx.vs.push_uniforms +
+            shader_remapped_sysval_offset(
+               vs, sysval_offset(graphics, vs.raw_vertex_offset));
+      }
+
+      struct panvk_precomp_ctx precomp_ctx = panvk_per_arch(precomp_cs)(cmdbuf);
+      enum panlib_barrier indirect_barrier =
+         PANLIB_BARRIER_JM_SUPPRESS_PREFETCH;
+      struct panlib_precomp_grid indirect_grid =
+         panlib_1d_with_jm_deps(1, 0, copy_desc_job_id);
+
+      if (draw->info.indirect.buffer_dev_addr != 0 && draw->info.index.size) {
+         const struct panlib_draw_indexed_indirect_helper_args args = {
+            .cmd = draw->info.indirect.buffer_dev_addr,
+            .index_buffer_ptr = cmdbuf->state.gfx.ib.dev_addr,
+            .index_size = draw->info.index.size,
+            .primitive_vertex_count = primitive_vertex_count(
+               translate_prim_topology(ia->primitive_topology)),
+            .primitive_restart = ia->primitive_restart_enable,
+            .varying_bufs_descs = draw->varying_bufs,
+            .varying_bufs_info = draw->indirect_info.varying_bufs,
+            .attrib_bufs_descs = draw->vs.attribute_bufs,
+            .attrib_bufs_infos = draw->indirect_info.attrib_bufs,
+            .attrib_bufs_valid = attrib_bufs_valid,
+            .attribs_valid = attribs_valid,
+            .attribs_descs = draw->vs.attributes,
+            .attribs_infos = draw->indirect_info.attribs,
+            .first_vertex_sysval = first_vertex_sysval,
+            .first_instance_sysval = first_instance_sysval,
+            .raw_vertex_offset_sysval = raw_vertex_offset_sysval,
+            .idvs_job = vs->info.vs.idvs ? draw->jobs.idvs.gpu : 0,
+            .vertex_job = draw->jobs.vertex.gpu,
+            .tiler_job = draw->jobs.tiler.gpu,
+         };
+         panlib_draw_indexed_indirect_helper_struct(&precomp_ctx, indirect_grid,
+                                                    indirect_barrier, args);
+      } else if (draw->info.indirect.buffer_dev_addr != 0) {
+         const struct panlib_draw_indirect_helper_args args = {
+            .cmd = draw->info.indirect.buffer_dev_addr,
+            .primitive_vertex_count = primitive_vertex_count(
+               translate_prim_topology(ia->primitive_topology)),
+            .varying_bufs_descs = draw->varying_bufs,
+            .varying_bufs_info = draw->indirect_info.varying_bufs,
+            .attrib_bufs_descs = draw->vs.attribute_bufs,
+            .attrib_bufs_infos = draw->indirect_info.attrib_bufs,
+            .attrib_bufs_valid = attrib_bufs_valid,
+            .attribs_valid = attribs_valid,
+            .attribs_descs = draw->vs.attributes,
+            .attribs_infos = draw->indirect_info.attribs,
+            .first_vertex_sysval = first_vertex_sysval,
+            .first_instance_sysval = first_instance_sysval,
+            .raw_vertex_offset_sysval = raw_vertex_offset_sysval,
+            .idvs_job = vs->info.vs.idvs ? draw->jobs.idvs.gpu : 0,
+            .vertex_job = draw->jobs.vertex.gpu,
+            .tiler_job = draw->jobs.tiler.gpu,
+         };
+         panlib_draw_indirect_helper_struct(&precomp_ctx, indirect_grid,
+                                            indirect_barrier, args);
+      } else {
+         const struct panlib_draw_indexed_helper_args args = {
+            .index_buffer_ptr = cmdbuf->state.gfx.ib.dev_addr,
+            .index_size = draw->info.index.size,
+            .first_index = draw->info.index.offset,
+            .index_count = draw->info.vertex.count,
+            .first_instance = draw->info.instance.base,
+            .instance_count = draw->info.instance.count,
+            .vertex_offset = draw->info.vertex.base,
+            .primitive_restart = ia->primitive_restart_enable,
+            .varying_bufs_descs = draw->varying_bufs,
+            .varying_bufs_info = draw->indirect_info.varying_bufs,
+            .attrib_bufs_descs = draw->vs.attribute_bufs,
+            .attrib_bufs_infos = draw->indirect_info.attrib_bufs,
+            .attrib_bufs_valid = attrib_bufs_valid,
+            .attribs_valid = attribs_valid,
+            .attribs_descs = draw->vs.attributes,
+            .attribs_infos = draw->indirect_info.attribs,
+            .first_vertex_sysval = first_vertex_sysval,
+            .first_instance_sysval = first_instance_sysval,
+            .raw_vertex_offset_sysval = raw_vertex_offset_sysval,
+            .idvs_job = vs->info.vs.idvs ? draw->jobs.idvs.gpu : 0,
+            .vertex_job = draw->jobs.vertex.gpu,
+            .tiler_job = draw->jobs.tiler.gpu,
+            .primitive_vertex_count = primitive_vertex_count(
+               translate_prim_topology(ia->primitive_topology)),
+         };
+         panlib_draw_indexed_helper_struct(&precomp_ctx, indirect_grid,
+                                           indirect_barrier, args);
+      }
+
+      /* Grab the index of the indirect helper job */
+      uint32_t prev_job = batch->vtc_jc.job_index;
+
+      if (vs->info.vs.idvs) {
+         pan_jc_add_job(&batch->vtc_jc, MALI_JOB_TYPE_INDEXED_VERTEX, false,
+                        false, 0, prev_job, &draw->jobs.idvs, false);
+      } else {
+         unsigned vjob_id =
+            pan_jc_add_job(&batch->vtc_jc, MALI_JOB_TYPE_VERTEX, false, true, 0,
+                           prev_job, &draw->jobs.vertex, false);
+
+         if (draw->jobs.tiler.gpu != 0) {
+            pan_jc_add_job(&batch->vtc_jc, MALI_JOB_TYPE_TILER, false, false,
+                           vjob_id, 0, &draw->jobs.tiler, false);
+         }
+      }
+   }
+
+   /*
+    * We split every ~1024 indirect draw.
+    * This is here for multiple reasons:
+    * - The indirect varying buffer offset need to be reset at some point to
+    * avoid going outside of bounds.
+    * - It is possible to always end up with timeouts for batches with 4k draws
+    * (see "dEQP-VK.api.command_buffers.many_indirect_draws_on_secondary") At
+    * the same time, because of how TLS works on Mali, we should not split too
+    * much as this will cause the TLS budget to go crazy.
+    */
+   if (batch->vtc_jc.job_index > (5 * 1024)) {
+      bool preload_fb =
+         cmdbuf->cur_batch && cmdbuf->cur_batch->vtc_jc.first_tiler;
+
+      panvk_per_arch(cmd_close_batch)(cmdbuf);
+
+      if (preload_fb)
+         panvk_per_arch(cmd_preload_fb_after_batch_split)(cmdbuf);
+
+      batch = panvk_per_arch(cmd_open_batch)(cmdbuf);
+      cmdbuf->state.gfx.vs.indirect_varying_bufs_infos = 0;
+   }
+
+   clear_dirty_after_draw(cmdbuf);
+   cmdbuf->state.gfx.vs.previous_draw_was_indirect = true;
+}
+
 static unsigned
 padded_vertex_count(struct panvk_cmd_buffer *cmdbuf, uint32_t vertex_count,
                     uint32_t instance_count)
@@ -1697,7 +1956,24 @@ panvk_per_arch(CmdDrawIndirect)(VkCommandBuffer commandBuffer, VkBuffer _buffer,
                                 VkDeviceSize offset, uint32_t drawCount,
                                 uint32_t stride)
 {
-   panvk_stub();
+   VK_FROM_HANDLE(panvk_cmd_buffer, cmdbuf, commandBuffer);
+   VK_FROM_HANDLE(panvk_buffer, buffer, _buffer);
+
+   if (drawCount == 0)
+      return;
+
+   /* We cannot support arbitrary draw count on JM */
+   assert(drawCount == 1);
+
+   struct panvk_draw_data draw = {
+      .info = {
+         .indirect.buffer_dev_addr = panvk_buffer_gpu_ptr(buffer, offset),
+         .indirect.draw_count = drawCount,
+         .indirect.stride = stride,
+      },
+   };
+
+   panvk_cmd_draw_indirect(cmdbuf, &draw);
 }
 
 VKAPI_ATTR void VKAPI_CALL
@@ -1705,7 +1981,25 @@ panvk_per_arch(CmdDrawIndexedIndirect)(VkCommandBuffer commandBuffer,
                                        VkBuffer _buffer, VkDeviceSize offset,
                                        uint32_t drawCount, uint32_t stride)
 {
-   panvk_stub();
+   VK_FROM_HANDLE(panvk_cmd_buffer, cmdbuf, commandBuffer);
+   VK_FROM_HANDLE(panvk_buffer, buffer, _buffer);
+
+   if (drawCount == 0)
+      return;
+
+   /* We cannot support arbitrary draw count on JM */
+   assert(drawCount == 1);
+
+   struct panvk_draw_data draw = {
+      .info = {
+         .index.size = cmdbuf->state.gfx.ib.index_size,
+         .indirect.buffer_dev_addr = panvk_buffer_gpu_ptr(buffer, offset),
+         .indirect.draw_count = drawCount,
+         .indirect.stride = stride,
+      },
+   };
+
+   panvk_cmd_draw_indirect(cmdbuf, &draw);
 }
 
 VKAPI_ATTR void VKAPI_CALL
-- 
GitLab

